Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebFormManager.cpp

```diff
@@ -0,0 +1,248 @@
+        "will-submit-form",
+        G_TYPE_FROM_CLASS(klass),
+        G_SIGNAL_RUN_LAST,
+        0, 0, nullptr,
+        g_cclosure_marshal_generic,
+        G_TYPE_NONE, 3,
+        JSC_TYPE_VALUE,
+        WEBKIT_TYPE_FRAME,
+        WEBKIT_TYPE_FRAME);
+}
+
+WebKitWebFormManager* webkitWebFormManagerCreate()
+{
+    return WEBKIT_WEB_FORM_MANAGER(g_object_new(WEBKIT_TYPE_WEB_FORM_MANAGER, nullptr));
+}
+
+void webkitWebFormManagerDidAssociateFormControls(WebKitWebFormManager* formManager, WebKitFrame* frame, Vector<GRefPtr<JSCValue>>&& elements)
+{
+    GRefPtr<GPtrArray> formElements = adoptGRef(g_ptr_array_sized_new(elements.size()));
+    for (const auto& element : elements)
+        g_ptr_array_add(formElements.get(), element.get());
+    g_signal_emit(formManager, signals[FORM_CONTROLS_ASSOCIATED], 0, frame, formElements.get());
+}
+
+void webkitWebFormManagerWillSendSubmitEvent(WebKitWebFormManager* formManager, GRefPtr<JSCValue>&& form, WebKitFrame* sourceFrame, WebKitFrame* targetFrame)
+{
+    g_signal_emit(formManager, signals[WILL_SEND_SUBMIT_EVENT], 0, form.get(), sourceFrame, targetFrame);
+}
+
+void webkitWebFormManagerWillSubmitForm(WebKitWebFormManager* formManager, GRefPtr<JSCValue>&& form, WebKitFrame* sourceFrame, WebKitFrame* targetFrame)
+{
+    g_signal_emit(formManager, signals[WILL_SUBMIT_FORM], 0, form.get(), sourceFrame, targetFrame);
+}
+
+static Node* nodeForJSCValue(JSCValue* value)
+{
+    auto* jsObject = JSValueToObject(jscContextGetJSContext(jsc_value_get_context(value)), jscValueGetJSValue(value), nullptr);
+    return jsObject ? JSNode::toWrapped(toJS(jsObject)->vm(), toJS(jsObject)) : nullptr;
+}
+
+/**
+ * webkit_web_form_manager_input_element_is_user_edited:
+ * @element: a #JSCValue
+ *
+ * Get whether @element is an HTML text input element that has been edited by a user action.
+ *
+ * Returns: %TRUE if @element is an HTML text input element tha has been edited by a user action,
+ *    or %FALSE otherwise
+ *
+ * Since: 2.40
+ */
+gboolean webkit_web_form_manager_input_element_is_user_edited(JSCValue* element)
+{
+    g_return_val_if_fail(JSC_IS_VALUE(element), FALSE);
+    g_return_val_if_fail(jsc_value_is_object(element), FALSE);
+
+    auto* node = nodeForJSCValue(element);
+    if (is<HTMLInputElement>(node))
+        return downcast<HTMLInputElement>(*node).lastChangeWasUserEdit();
+
+    if (is<HTMLTextAreaElement>(node))
+        return downcast<HTMLTextAreaElement>(*node).lastChangeWasUserEdit();
+
+    return FALSE;
+}
+
+/**
+ * webkit_web_form_manager_input_element_auto_fill:
+ * @element: a #JSCValue
+ * @value: the text to set
+ *
+ * Set the value of an HTML input element as if it had been edited by
+ * the user, triggering a change event, and set it as filled automatically.
+ * If @element is not an HTML input element this function does nothing.
+ *
+ * Since: 2.40
+ */
+void webkit_web_form_manager_input_element_auto_fill(JSCValue* element, const char* value)
```

Comment:
I'll be eager to test if this actually works. Currently I have to type and delete a character to get many websites to accept my autofilled usernames/passwords.